### DISCUSSION TITLE
Welcome to Discussions!

### DISCUSSION THREAD

**Discussion by morpho-matters** [morpho-matters/morpholib#3] [Announcements]
## 👋 Welcome!
  You can use this space to ask questions about Morpho and share ideas.

**Comment by albert-smi** [comment]
Hi @morpho-matters,
As a project of mine i'd also like to creare a library to animate math and i found your work pretty inspiring. So I wanted to ask you what tutorial or guide you used to create this library. How did you create the entire structure? Did you find any step-by-step guides that explained how to create it, at least initially?

**Reply by morpho-matters** [reply]
Hey!

I actually didn't follow any kind of guide or tutorial to make this library because I originally never intended for it to be this big or complex. Most of the library's design was informed from the kind of tool I wanted to have, particular animation problems I wanted or needed to solve, and frictions I encountered when using the tool in its earlier iterations. It was very much a thing that just grew out organically, originally just as an unserious side project. I only had a loose idea of the overall structure in the beginning and I simply made changes and additions along the way that helped the library feel nicer for me to use and more capable.

I realize that may not be the most helpful answer for you in kickstarting your own project, but I will say a lot of the initial design ideas I had for Morpho came from traditional animation techniques, and software such as Adobe Flash/Animate. I knew I wanted the workflow to feel akin to that, where you set keyframes and have the software tween between them, as opposed to the more "PowerPoint"-like approach, where you apply pre-made animations/effects to objects. I also knew I wanted a sort of "minimalist" design to it where you could do a lot with a few basic tools/concepts once you learned how to use them effectively, as opposed to having a gigantic catalog of pre-made functions that do highly specific things. I think having a clear idea of how I wanted the tool to look and feel helped with getting the project off the ground and to maintain interest in it in the long run.

Anyway, I hope that helps some! Also happy to go into more detail about aspects of Morpho's design if you have further questions about it!

**Reply by albert-smi** [reply]
Ok @morpho-matters, well i'm currently trying to test both manim and morpholib, to see the differences. Is there a way to have an effect where the letters of a text become the letters of another and same for shapes? I literally mean like transform in manim. Is there a Transform-like effect in morpholib?

**Reply by morpho-matters** [reply]
There's a few different ways to achieve a manim-like Transform effect for text. The cheapest way is this one:
```python
import morpholib as mo
mo.importAll()

mainlayer = mo.Layer()
mation = mo.Animation(mainlayer)

text = mainlayer.Actor(mo.text.MultiPText("Hello"))
text.newendkey(30).text = "World"

mation.play()
```
but a more manim-like Transform can be achieved by converting the text into a spline first:
```python
import morpholib as mo
mo.importAll()

mainlayer = mo.Layer()
mation = mo.Animation(mainlayer)

text = mainlayer.Actor(mo.text.MultiPText("Hello").toSpline())
text.newendkey(30, mo.text.MultiPText("World").toSpline())

mation.play()
```
Finally, there's a way to transform LaTeX as well, which is what I most commonly morph in my own animations (note to run this code you will need to have a LaTeX distribution installed on your machine):
```python
import morpholib as mo
mo.importAll()

mainlayer = mo.Layer()
mation = mo.Animation(mainlayer)

equ = mainlayer.Actor(mo.latex.parse(r"E = mc^2", boxWidth=8, fill=[1,1,1]))
equ.newendkey(30).replaceTex(r"a^2 + b^2 = c^2", boxWidth=8, fill=[1,1,1])

mation.play()
```
Generally, the way transforming/morphing works in Morpho is you set a starting and ending state for an actor (e.g. text, LaTeX, path, etc.) and Morpho will automatically smoothly transition from the starting state to the ending state during animation playback. You don't ever explicitly declare you want to transform, you just change an actor's state at a different point in time and it will auto-transform it.
